docs/companion.md

```diff
@@ -3,3 +3,705 @@ sidebar_position: 4
+
+#### `s3`
+
+Companion comes with signature endpoints for AWS S3.
+These can be used by the Uppy client to sign requests to upload files directly to S3, without exposing secret S3 keys in the browser.
+Companion also supports uploading files from providers like Dropbox and Instagram directly into S3.
+
+##### `s3.key` `COMPANION_AWS_KEY`
+
+The S3 access key ID.
+
+##### `s3.secret` `COMPANION_AWS_SECRET`
+
+The S3 secret access key.
+
+##### `s3.bucket` `COMPANION_AWS_BUCKET`
+
+The name of the bucket to store uploaded files in.
+
+##### `s3.region` `COMPANION_AWS_REGION`
+
+The datacenter region where the target bucket is located.
+
+##### `s3.awsClientOptions`
+
+You can supply any [S3 option supported by the AWS SDK](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#constructor-property)
+in the `providerOptions.s3.awsClientOptions` object, _except for_ the below:
+
+* `accessKeyId`. Instead, use the `providerOptions.s3.key` property.
+  This is to make configuration names consistent between different Companion features.
+* `secretAccessKey`. Instead, use the `providerOptions.s3.secret` property.
+  This is to make configuration names consistent between different Companion features.
+
+Be aware that some options may cause wrong behaviour if they conflict with Companion’s assumptions.
+If you find that a particular option does not work as expected, please [open an issue on the Uppy repository](https://github.com/transloadit/uppy/issues/new) so we can document it here.
+
+##### `s3.getKey(req, filename, metadata)`
+
+Get the key name for a file. The key is the file path to which the file will be uploaded in your bucket.
+This option should be a function receiving three arguments:
+
+* `req`, the HTTP request, for _regular_ S3 uploads using the `@uppy/aws-s3` plugin.
+  This parameter is _not_ available for multipart uploads using the `@uppy/aws-s3-multipart` plugin;
+* `filename`, the original name of the uploaded file;
+* `metadata`, user-provided metadata for the file.
+  See the [`@uppy/aws-s3`](https://uppy.io/docs/aws-s3/#metaFields) docs. The `@uppy/aws-s3-multipart` plugin unconditionally sends all metadata fields, so they all are available here.
+
+This function should return a string `key`.
+The `req` parameter can be used to upload to a user-specific folder in your bucket, for example:
+
+```js
+app.use(authenticationMiddleware)
+app.use(uppy.app({
+  providerOptions: {
+    s3: {
+      getKey: (req, filename, metadata) => `${req.user.id}/${filename}`,
+      /* auth options */
+    },
+  },
+}))
+```
+
+The default implementation returns the `filename`,
+so all files will be uploaded to the root of the bucket as their original file name.
+
+```js
+app.use(uppy.app({
+  providerOptions: {
+    s3: {
+      getKey: (req, filename, metadata) => filename,
+    },
+  },
+}))
+```
+
+#### `COMPANION_AWS_SECRET_FILE`
```

Comment:
@mifi we need more info on how this is used and what the file should look like. Can you make a suggestion?

Reply:
seems like it's not in use so i'll remove it

Reply:
So this should be a relative path to a file with any extension which _only_ contains the secret directly, nothing else?